src/uuid.cr

```diff
@@ -107,16 +107,59 @@ struct UUID
     new(bytes, variant, version)
   end
 
+  def self.parse?(value : String, variant = nil, version = nil)
```

Comment:
It would be better to create an error yielding method and use that in `.parse?` and `.new` instead of duplicating the code.

Reply:
@Sija It was the first thing I tried and it was, in fact, worse. The block passed in from `parse?` was nice because it was empty, but that wasn’t true for the block passed from `.new`.

Every single failure mode (that is, every single exception that can be raised in `.new` and `.hex_pair_at` in the current version) had to be accounted for in the block passed in from `.new` (and then passed into `hex_pair_at`) as well as the method actually performing the parsing. So the block had to perform all the same checks a second time to figure out _which_ failure mode had been tripped, which not only duplicates the code anyway, but also is slower.

Reply:
I don't follow. I meant to create a parse method which would return the error instead of raising, so the caller could decide whether he want to raise or ignore it.

Reply:
Do you have an example of the behavior you’re talking about? Because you said an “error yielding method” above, which I understood to be something like `Hash#fetch` where the block is yielded on a miss. And what you’re saying here sounds like something else. A code example of what you mean might help clarify.

Reply:
You're right. My first post was misguided, I meant a method _returning_ error.

Reply:
Interesting, so it you're envisioning a method that returns kinda like how you'd do it in Go, but with a Crystal-style flavor?

```crystal
private def parse_string(value : String, variant = nil, version = nil) : UUID | ArgumentError
  def self.new(value : String, variant = nil, version = nil)
    bytes = uninitialized UInt8[16]
    case value.size
    when 36 # Hyphenated
      {8, 13, 18, 23}.each do |offset|
        if value[offset] != '-'
          return ArgumentError.new "Invalid UUID string format, expected hyphen at char #{offset}"
        end
      end
      {0, 2, 4, 6, 9, 11, 14, 16, 19, 21, 24, 26, 28, 30, 32, 34}.each_with_index do |offset, i|
        if hex = hex_pair_at? value, offset
           bytes[i] = hex
         else
           return ArgumentError.new("Invalid hex character at position #{i * 2} or #{i * 2 + 1}, expected '0' to '9', 'a' to 'f' or 'A' to 'F'")
         end
      end
    when 32 # Hexstring
      16.times do |i|
        bytes[i] = hex_pair_at value, i * 2
      end
    when 45 # URN
      return ArgumentError.new "Invalid URN UUID format, expected string starting with \"urn:uuid:\"" unless value.starts_with? "urn:uuid:"
      {9, 11, 13, 15, 18, 20, 23, 25, 28, 30, 33, 35, 37, 39, 41, 43}.each_with_index do |offset, i|
        if hex = hex_pair_at? value, offset
           bytes[i] = hex
         else
           return ArgumentError.new("Invalid hex character at position #{i * 2} or #{i * 2 + 1}, expected '0' to '9', 'a' to 'f' or 'A' to 'F'")
         end
      end
    else
      return ArgumentError.new "Invalid string length #{value.size} for UUID, expected 32 (hexstring), 36 (hyphenated) or 45 (urn)"
    end
    new(bytes, variant, version)
  end
end
```

And then the other methods would be something like this?

```crystal
def self.new(value, variant = nil, version = nil) : UUID
  case result = parse_string(value, variant, version)
  when UUID
    result
  when Exception
    raise result
  end
end

def self.parse?(value, variant = nil, version = nil) : UUID?
  if (uuid = parse_string(value, variant, version)).is_a? UUID
    uuid
  end
end
```

Reply:
yup, that was the idea :)

Reply:
The return error strategy still allocates an exception instance and builds the error messages- which is just useless if it's going to be thrown out anyway.
I think a better option to avoid duplication would be an internal method with a parameter that determines whether to raise or not.
Not conviced that's really necessary here. A bit duplication isn't bad. And the duplicated code is quite simple. I'm happy to accept the current implementation.

Reply:
Maybe it could be an option to separate input verification from the actual parsing. Then we'd only duplicate the former and share the latter.

Simplified example:
```cr
def self.parse?(value)
  return nil if value.size == 45 && !value.starts_with? "urn:uuid:"

  parse_internal(value) { return nil }
end
def self.new(value)
  raise ArgumentError.new if value.size == 45 && !value.starts_with? "urn:uuid:"

  parse_internal(value) { |pos| raise ArgumentError.new "Invalid hex character at #{pos}" }
end
private def self.parse_internal(value)
  # ...
  parse_hex_pair?(value, i) || yield i
end
```

Reply:
IMO throwing out **one** allocation doesn't sound that bad in light of having a cleaner impl. and giving the user precise errors in case of failure.